### DISCUSSION TITLE
Next.js's fetch API extension seems to mess with streaming

### DISCUSSION THREAD

**Discussion by klcantrell** [vercel/next.js#68513] [Help]
### Summary

Consider the following fetch call and stream handling running on the server:

```typescript
const response = await fetch("http://backend:8000");

if (!response.ok || !response.body) {
  return;
}

const reader = response.body
  .pipeThrough(new TextDecoderStream("utf-8"))
  .getReader();

try {
  while (true) {
    const { done, value: chunk } = await reader.read();
    if (done) {
      break;
    }

    console.log(chunk);
  }
} finally {
  reader.releaseLock();
}
```

Instead of multiple chunks that start to arrive right away and come in over time, I've run into an issue where the entire payload arrives (eventually) all at once in one giant chunk.

With just this tweak, I get the behavior I want:

```typescript
const response = await fetch("http://backend:8000", { cache: "no-cache" }); // included `{ cache: "no-cache" }` setting
```

Can anyone help me understand if this is a bug or intended behavior? I'm having trouble finding documentation or prior discussions about this.

Also, I ran the same code in just a plain Node script and the response was streamed as expected regardless of the `cache` setting (tried it with default settings as well as "force-cache" and "no-cache"). This makes me wonder if Next.js's fetch API extension is doing something to prevent streaming, and I wonder if that's intentional.

### Additional information

```js
`next info`

Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:13:18 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6030
  Available memory (MB): 36864
  Available CPU cores: 12
Binaries:
  Node: 20.11.1
  npm: 10.2.4
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.5 // Latest available version is detected (14.2.5).
  eslint-config-next: 14.2.5
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.5.4
Next.js Config:
  output: N/A
```

### Example

https://github.com/klcantrell/nextjs-with-langchain

**Comment by klcantrell** [comment]
One more piece of the puzzle. The issue described only happens on the server in Next.js. The streaming behavior works as expected when the `fetch` call is done client-side even without `{ cache: "no-cache" }` specified.

**Comment by brunolm** [comment]
It does.

If you're using AI stuff, you might want to consider getting the SDK.

One other alternative is potentially installing node-fetch? I'm not sure if that works, but maybe.
